schulcloud/FilesViewController.swift

```diff
@@ -33,16 +34,58 @@ class FilesViewController: UITableViewController, NSFetchedResultsControllerDele
     }
     
     @IBAction func didTriggerRefresh() {
-        FileHelper.updateDatabase(forFolder: currentFolder)
-            .onSuccess { _ in
+        
+        let successBlock : () -> () = {
+            DispatchQueue.main.async {
                 self.performFetch()
             }
-            .onFailure { error in
-                log.error(error)
-            }
-            .onComplete { _ in
+        }
+        let completeBlock : () -> () = {
+            DispatchQueue.main.async {
                 self.refreshControl?.endRefreshing()
             }
+        }
+
+        if FileHelper.coursesDirectoryID == currentFolder.id {
+            CourseHelper.fetchFromServer()
+            .andThen { (result) in
+                if let changes = result.value {
+                    FileHelper.process(changes: changes, inFolder: self.currentFolder, managedObjectContext: managedObjectContext)
+                }
+            }.onSuccess(callback: { (_) in
+                successBlock()
+            })
+            .onFailure(callback: { (error) in
+                print("Failure: \(error)")
+            }).onComplete(callback: { (_) in
+                completeBlock()
+            })
+        } else if FileHelper.sharedDirectoryID == currentFolder.id {
+            fileSync.sharedDownload()
+            .map { (objects) -> Void in
+                for json in objects {
+                    FileHelper.updateDatabase(contentsOf: self.currentFolder, using: json)
+                }
+                return ()
+            }
+            .onSuccess(callback: successBlock)
+            .onFailure(callback: { (error) in
+                print("failure: \(error)")
+            }).onComplete(callback: { (_) in
+                completeBlock()
+            })
+        } else {
+            fileSync.downloadContent(for: currentFolder)
+            .map { json -> () in
+                FileHelper.updateDatabase(contentsOf: self.currentFolder, using: json)
+                return ()
+            }.onSuccess(callback: successBlock)
+            .onFailure { (error) in
+                log.error(error)
+            }.onComplete { (_) in
+                completeBlock()
+            }
+        }
```

Comment:
DRY. You have the calls `onSuccess`, `onFailure` and `onComplete` three times. Just create a variable for the future before the if/elseif/else block and do the future handling once afterwards.

Reply:
Each Futures are of different type. I will see if I can factor it, but unsure if it will work

Reply:
I have experimented with some solution for fixing that, maping all Futures to a `Future<Any, SCError>` to bind them to a variable and then chain the success/failure/complete. I think DRY does not make things more readable, complicate things for no value. As much as i agree that repeating feels wrong, it is a superior solution than any alternative in this context.

Reply:
I would add that the meat of the problem isn't the repeating part, more that we have 2 special cases to handle, for courses directories and shared directory.